app/controllers/api/v1/sites_controller.rb

```
@@ -3,7 +3,79 @@
+      end
+
+      # rubocop:disable Metrics/LineLength
+      # rubocop:disable Metrics/CyclomaticComplexity
+      def set_widget_config
+        return render json: { errors: ['You must be an admin to do this.'] }, status: :forbidden unless current_user.admin?
+
+        return render json: { errors: ['No parameters provided, useless invocation.'] }, status: :bad_request if allowed_config.empty?
+
+        return render json: { errors: ['You must provide a valid position.'] }, status: :bad_request unless !allowed_config[:position].present? || %w[top-right top-left bottom-right bottom-left].include?(allowed_config[:position])
+
+        site = current_user.site
+        site.transaction(isolation: :serializable) do
+          site.lock!
+          site.update!(allowed_config.permit(:title, :position, :subtitle))
+
+          encoded_main_icon = allowed_config[:main_icon]
+          encoded_launcher_icon = allowed_config[:launcher_icon]
+
+          if encoded_main_icon.present?
+            encoded_main_icon = encoded_main_icon.partition(',').last
+            main_icon = MiniMagick::Image.read(Base64.decode64(encoded_main_icon))
+            file_name = "main_icon#{site.id}.#{main_icon.type.downcase}"
+            main_icon.write file_name
+            File.open(file_name) do |file|
+              site.main_icon = file
+            end
+            main_icon.destroy!
+            File.delete(file_name)
+          end
+
+          if encoded_launcher_icon.present?
+            encoded_launcher_icon = encoded_launcher_icon.partition(',').last
+            launcher_icon = MiniMagick::Image.read(Base64.decode64(encoded_launcher_icon))
+            file_name = "launcher_icon#{site.id}.#{launcher_icon.type.downcase}"
+            launcher_icon.write file_name
+            File.open(file_name) do |file|
+              site.launcher_icon = file
+            end
+            launcher_icon.destroy!
+            File.delete(file_name)
+          end
+          site.save!
```

The exclamation marks in create!, save!, update! tells ruby to throw an exception if something goes wrong.
But there is no exception handling here so there is no point in using this syntax. We can Simply use save, create, update.

---

You cannot, the exception is not handled because, if raised, it would indicate a concurrency problem generated at database level, the exceptions raised are not handled because there is nothing to do with them. Upon being raised the function must abort, but it is important for them to be raised, a exception raising from within a transaction block will generate the complete transaction to rollback returning the database to a consistent state, it will also generate the transaction to re raise the exception aborting the whole function with an internal server error return status